cache/cache.go

```diff
@@ -0,0 +1,130 @@
+// Package cache implements a fixed size circular cache which can be used to
+// cache a fixed number of items. Once the underlying array is full, the oldest
+// item in the cache will be overwritten. The cache thread-safe.
+package cache
+
+import (
+	"errors"
+	"sync"
+)
+
+// noIndex is the value of the cache's index when no values have been added.
+const noIndex = -1
+
+// errInvalidSize is returned when an invalid size for a cache is provided.
+var errInvalidSize = errors.New("cache size must be > 0")
+
+// Circular is a cache which retains a set of values in memory, and
+// overwrites the oldest item in the cache when a new item needs to be
+// written.
+type Circular struct {
+	// size is the total number of items that the cache will hold.
+	size int
+
+	// index is the index of the most recently written item in the
+	// cache.
+	index int
+
+	// total is the total number of items that have been added to the
+	// cache.
+	total int
+
+	// items is the set of cached items.
+	items []interface{}
+
+	sync.RWMutex
+}
+
+// NewCircular returns a new circular cache with the size provided. It will fail
+// if a zero of negative size parameter is provided.
+func NewCircular(size int) (*Circular, error) {
+	if size <= 0 {
+		return nil, errInvalidSize
+	}
+
+	return &Circular{
+		index: noIndex,
+		total: 0,
+		size:  size,
+
+		// Create a slice with length and capacity equal to the size of
+		// the cache so that we do not need to resize the underlying
+		// array when we add items.
+		items: make([]interface{}, size),
+	}, nil
+}
+
+// Add adds an item to the cache, overwriting the latest item if the cache
+// is full.
+func (c *Circular) Add(item interface{}) {
+	c.Lock()
+	defer c.Unlock()
+
+	// Increment the index and wrap by size if necessary.
+	c.index = (c.index + 1) % c.size
+
+	// Set the item in the next free index in the items array.
+	c.items[c.index] = item
+
+	// Increment the total number of items that we have cached.
+	c.total++
+}
+
+// List returns a copy of the items in the cache ordered from the oldest to
+// newest item.
+func (c *Circular) List() []interface{} {
+	c.RLock()
+	defer c.RUnlock()
+
+	switch {
+	// If no items have been cached yet, we can just return a nil list.
+	case c.index == noIndex:
+		return nil
+
+	// If we have added fewer items than the cache size, we can simply
+	// return the total number of items from the beginning of the list
+	// to the total. This special case is added because the oldest item
+	// is at the beginning of the underlying array, not after the index,
+	// when we have not filled the array yet.
+	case c.total < c.size:
```

Comment:
Not sure I'm following this comment? 
The logic after this doesn't work for an array that isn't full yet, it'll return things in the wrong order. Theis covers the special case where the oldest item is at index 0, rather than being after the index.

Reply:
Sorry, misunderstood something and made an incorrect comment. All is fine here.